### PR TITLE
feat: allow custom naming of events and event properties

### PR DESCRIPTION
**Pull request by alexkim205**
## Problem

Users have asked for custom naming of event and event property definitions.

## Changes

👉 *Stay up-to-date with our [coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Comment by mariusandra** [comment]
I don't know the backstory here, but I'm hesitating if this is something we should do. It's always easy to add more things... and then we basically need to support them forever.

What's the scope here? We allow users to rename "$pageview" to "This Is My Page That I Viewed", and show that in all places, including the property filter, autocomplete, reports, insights, labels, etc?

CC @paolodamico @marcushyett-ph

**Comment by marcushyett-ph** [comment]
@mariusandra This seems like a fairly reasonable ask, especially for scenarios where people are trying to avoid confusion that was only realized after the event was ingested:
* "I spelled the event name wrong initially and I need to update it"

Even in the demo of one of our competitors is this use-case highlighted (Updating the display name for a grammatically confusing event):
image

On the scope question, if we don't use the same display name everywhere - it will be even more confusing, so we have to be pretty exhaustive here.

@mariusandra can you share any more detail on your concerns around the cost of supporting this going forward?

There's a risk that people will rely on updating display names to mask bad instrumentation.

@alexkim205 it'd also be great if you could link to the relevant customer feedback?

**Comment by neilkakkar** [comment]
This seems a bit confusing, when combined with the custom naming on a funnel / trend.

We probably wouldn't want users to use this way to do things when all they want is to clarify what Pageview with filter on the second step of this funnel is for.. And I get a bit apprehensive when we introduce more ways to do the same thing (albeit being the wrong way 🤷 ).

Also keen to hear about user requests that came in around this!

**Comment by mariusandra** [comment]
@marcushyett-ph it's more of a gut feeling that we're increasing complexity more than we realise, but few specific things come to mind:

- Confusion due to inconsistency. I'm sure someone will bang their head against a wall looking for the "booking completed" event in their system, only to later realise it's actually a rename of "booking finished".
- The autocomplete in the taxonomic property filter will need to accommodate for all the names, and searching for "finished" and only seeing "booking completed" will surely confuse someone.
- What about existing posthog "humanized" events like "$pageview" -> "🦔 Pageview". Can you rename them? Are these custom names?
- As Neil pointed out, we already have a way to rename trend and funnel lines. Adding another rename in a different place could confuse people.
- We will never have all events with all names cached in memory, so if you load an insight with the "booking completed" event, we'll need to make another async request to fetch the right name "booking finished". When you open an insight with some filters from a link, the page will either show a spinner when we try to make sure nobody has renamed any event, or then we might start flicking names at random. We could "hardcode" the "custom name" into the filters, but what if it changes again?
- What if you have events "a" and "b", and rename "a" -> "b" and "b" -> "a"? Extreme case, I know, but you can start to mask real data somehow.

I'm not saying not to do this... and we already somewhat do this with the "$pageview" -> "🦔 Pageview" events. Just let's be sure we really want this.

**Comment by marcushyett-ph** [comment]
@mariusandra it sounds like this is a change we might regret if we get it wrong, so we should have a higher bar for getting it right, and/or reduce the scope.

I don't see an obvious use-case to allow people to rename $pageview etc. So it would be reasonable to exclude these from the scope. But the other edge cases you mention seem like we should get some customer feedback on (e.g. renaming to a previously used name, overlap with other renaming features, etc.)

I'll let @alexkim205 chime in with the customer context here, I'd particularly like to understand the difference for them between this and just renaming in the insight itself?

I'm keen to help follow up with the customers too to get the answers here so we can make the call quickly.

**Comment by alexkim205** [comment]
Appreciate the conversation here. This initiative came out of this Slack User [thread](https://posthogusers.slack.com/archives/C01GLBKHKQT/p1645535282257119) where a user wanted a way to change the display name for custom events (i.e., "Product Category" instead of product_category).

The user said that this need comes from the product team being able to see friendlier names in the taxonomic filters (not already built insights), which will help them discover and use the correct events while they're building insights. I see insight filter renaming and event display names as two separate features to tackle, and it might be worth getting @clarkus's opinion here about how we would differentiate this in the UI so that those two don't clash in the same context.

> ...we should have a higher bar for getting it right, and/or reduce the scope.

100%. I think a prerequisite for this PR would be to compile everywhere in the app we show the `{event|property}.name` which might give us a better pict

> What about existing posthog "humanized" events like "$pageview" -> "🦔 Pageview". Can you rename them? Are these custom names?

I agree with both of yall. I don't think any PH managed events/properties should be renamed. For now I've disabled this custom_name field for all PH events.

> We will never have all events with all names cached in memory, so if you load an insight with the "booking completed" event, we'll need to make another async request to fetch the right name "booking finished". When you open an insight with some filters from a link, the page will either show a spinner when we try to make sure nobody has renamed any event, or then we might start flicking names at random. We could "hardcode" the "custom name" into the filters, but what if it changes again?

You bring up a really good point and it's been something that's been keeping me up at night 🛌 . I think there're two ways we can go here: (1) The `{event|property}DefinitionsModel` that I think you're talking about where we're caching this metadata in memory (which has its downsides). Not sure if we're also planning to deprecate these model logics. The one that I like is (2) if we dynamically fetching these event and property display names in the backend on every query that fetches information about an event/property. The clear downside here would be an extra query on each request but I bet a cache here would help. Thoughts?

### Some thoughts

Initially I was thinking custom display names would apply to properties, but I think it makes sense to also leave that out of scope as well since properties tend to be quite overloaded across events.

Technically, wouldn't an action composed of a single custom event solve for this use case? It would mean that the team will need to create a action for every event they have, but they'd essentially be doing the same amount of work for event definitions. This is interesting since a lot of taxonomy work has been bringing event definitions up to speed with actions (tags, descriptions), and this falls under that same category of work. I think the question here is whether this level of complexity is worth it or if this use case can be solved with today's actions and better in-app documentation. @marcushyett-ph I'd be curious to know if actions solve most user's use cases here? I'd also be happy to jump on some user calls/threads with you.

**Comment by clarkus** [comment]
This is the kind of stuff that needs ownership within the project teams using PostHog, so my gut says we should enable that. If it's a free-for-all and there isn't gated access to this kind of management task, then I agree with @mariusandra's concerns about complexity and understanding the source of truth when picking events. It's kind of like tagging - it needs a strategy or ownership to establish structure and drive consistent usage. If events were able to be renamed, we'd possibly see less renaming of query steps. I do think there is a case for both, though. Renaming a query step takes into account more than the base event (filters, aggregation methods, etc), so it makes sense that a user might want to rename the step to encompass the full definition of that parameter. 

@mariusandra's list of concerns could be partially addressed with the design changes. For example, when names change we can show both the friendly name and the raw name at different priorities. We already do this for renaming query steps, so we could take the same approach here. With the new taxonomy definitions cards, it'll be a lot easier to communicate this much detail without displacing the current UI. For autocomplete, can we match against both raw and friendly names? Over time, I think the friendly names will win out as more users use them. 

TL;DR - I think we should do this, but agree we should aim for a high quality solution that inspires trust. That will likely mean some user testing with users who reported this problem, and a few iterations to ensure we understand the constraints enough to inform the design. Happy to help with ideas when y'all are ready for that step.

**Comment by paolodamico** [comment]
Seems like a great opportunity to try a medium-sized RFC? With product specs we can flush out all the edge cases and address all these concerns.

**Comment by mariusandra** [comment]
> The user said that this need comes from the product team being able to see friendlier names in the taxonomic filters (not already built insights), which will help them discover and use the correct events while they're building insights.

I guess changing the event "insight viewed" to "insight viewed - USE THIS ONE NOT THE OTHER ONE" will help discovery, but so might other things like starring, tagging, etc. 

Other than worries about accidental inconsistency and accidental complexity adding this brings, I'm not against the feature. Just let's be sure we know what all the implications are, and if any workaround can help these users, before committing to a feature we won't be able to remove once we build it.

> Initially I was thinking custom display names would apply to properties, but I think it makes sense to also leave that out of scope as well since properties tend to be quite overloaded across events.

Definitely leave properties out of v1. Getting this right with a high enough bar for quality for events will be complicated enough that let's not make our lives harder than they should be :).

> Technically, wouldn't an action composed of a single custom event solve for this use case?

Indeed, could "just use actions" be a suitable workaround for the time being?

> Renaming a query step takes into account more than the base event 

Yup, renaming lines on trends is a separate feature. I'm sure we'll be able to explain in the interface the difference rather well.

> Seems like a great opportunity to try a medium-sized RFC? 

Seems so!

**Comment by marcushyett-ph** [comment]
Great discussion. My proposed next steps:

1. @alexkim205  - Kick off a [medium sized RFC](https://github.com/PostHog/product-internal/pull/287/files) - I can jump in and fill out any missing context
2. @marcushyett-ph set up customer interview ASAP with the user who reported this - run through their scenarios to help inform decision in RFC (also bounce the idea of using actions off them) @alexkim205 are there any other folks who reported this?
3. @marcushyett-ph aggregate feedback from Customer Success Team and other threads to inform the RFC

Once we have this, make the decision on what we ship build and how we handle the technical and design risks.

**Comment by marcushyett-ph** [comment]
I'm getting a bunch of calls booked in for this for early next week. @alexkim205  will make sure you're on the invite, but others are welcome to join too if keen @clarkus?

Proposed structure for the calls:
1. Get background on their general approach to data management, who does it, what's the objective, what problems do they need to solve with it?
2. Demonstrate what we've already built and capture feedback for future iterations
3. Deep dive on renaming events? Understand the use-cases and motivations for renaming them / edge-cases? Benefits? whether they do or would use actions for this today? How it differs from renaming on an insights?

**Comment by timgl** [comment]
Flyby comment: A logical follow up that I'm expecting customers will ask for next is "can I combine the event I misspelled and the correct event together into a single event"? This usecase is covered perfectly by Actions, but will be a nightmare if we go down the current route.

I suggest:
- Rename actions to 'custom events' or something that will make an action feel more like a real event.
- Rethinking how we show actions in the events dropdown. Maybe in the same list?
- If people try renaming an event, instead create an action and hide the current event?

Again we can make the frontend look like anything we want, but this should reduce the complexity that Marius is worried about

**Comment by clarkus** [comment]
Actions could be a great solution. We've already talked about combining the events and actions list, so that could be a good next step. I also like the idea of renaming the feature to something a bit more obvious. "Custom events" totally works, but might conflict with our use of the same term on the paths insight event sources.

**Comment by alexkim205** [comment]
@rcmarron called this "calculated events" which I thought was a really good term for it.

**Comment by timgl** [comment]
Yep calculated events was the term I was looking for!

**Comment by marcushyett-ph** [comment]
@alexkim205 and I have spoken a couple customers so far (including the person who originally asked for this) - we have another booked in next week... this is not the final conclusion but some interim context that might help is tighten the scope:

- Customers want to use data management standardize how analytics is done, to improve consistency and accuracy
- Actions are currently used to group similar events and or give them a friendly “display name” (but inconsistently)
- There is confusion over the difference between Actions and Events (particularly for new users)
- If Actions are used to create display names for events users should be able to hide the original events to avoid duplication and confusion over which to use
- Renaming the series on a chart is very separate from renaming the underlying event, the same event may be “presented” in different ways in a chart (e.g. pageview filtered by url = home, may be presented as “homepage views”)
- People have no strong preference for or against renaming posthog events like pageview but gave no examples of when they would do this
- Instrumentation is always evolving, actions enable customers to keep a consistent query experience even as the underlying events change
- Customers want people to have a consistent, distinct and limited vocabulary for important events - so everyone knows what they mean, how to use them and how important they are (e.g. Download [means the same thing, no matter what device it was downloaded to])

**Comment by marcushyett-ph** [comment]
@alexkim205 and I have spoken to a load of customers now, I'll write up the conclusions and share the raw findings very soon. 

TL;DR: The feedback on naming events has centered around using Actions to name, combine and split events, suggesting that this current approach is not likely to solve for users needs, so we can probably close this PR.
